docs/PERSISTENCE.md

```diff
@@ -0,0 +1,57 @@
+- Create: The log stream receives no inputs. It will first create one plog: the active plog. Then, it will record the active plog id and mark this log stream to be ready to handle append and read operations.
+- Append offset 
+- Append without offset: This operation takes one input: `payload`. It will return the plog id to indicate which plog stores this payload, and the current offset of this plog. The log stream will first check whether the current plog will excess the size limit if it appends the payload to the current plog. If it will not exceed the size limit, the log stream client will append the input payload to the corresponding plog and return success. Otherwise, it will first seal the old plog, record the sealed offset of the old plog, create a new plog, record the plog id of the new plog, then write the payload to the new plog. We call this process `Plog Switch`. During this process, all the upcoming write should wait until the switching process finished in order to ensure correctness.
+- Read: This operation takes 3 inputs: `start plog id`, `start offset` and `size`. It is possible that the input size is too large so that the log stream will read the contents from multiple plogs. The log stream will first check how many plogs this read operation will involve by checking the vector of plogs it used. Then, it will read the requested contents from all the involved plogs asynchronously. Once it receives the response from all the plogs, it will return a vector of payloads.
+
+## Optimization
+When we switch the old plog to the new plog for the write operation, all the upcoming write requests will be blocked until the switching process finished, which is very inefficient. We used several optimization methods to improve the performance of this write operation:
+- Pre-allocated Plog: When we create the log stream, we will not only create one active plog. Instead, we will also create a pre-allocated next plog which is used to perform plog switch faster. Therefore, when we do the plog switch, we can use the pre-allocated plog as the new plog immediately, instead of waiting for creating a new plog. Then, we can create the new pre-allocated plog in the background afterward.
+- Make the plog switch process to be asynchronous: When we switch the old plog to the new plog, it is not necessary for us to block the upcoming write operations until the switching process finished. Instead, We can perform the writes against the new plog without waiting, but not acknowledge them until we record this switch. 
+
+# Metadata Manager
+The metadata manager is a subservice that manages the metadata of all types of data services in the same partition. Each partition will only have one metadata manager. The metadata manager holds its own log stream: metadata log stream, to store the metadata of all data services. However, we need to record the metadata of metadata log stream as well. We store it into CPO.
```

Comment:
A bit confusing here. I think the metadata service in reality is our CPO, and there is just one service per cluster, not per partition